src/responder/common/cache_req/cache_req.c

```diff
@@ -0,0 +1,558 @@
+/*
+    Authors:
+        Pavel Březina <pbrezina@redhat.com>
+
+    Copyright (C) 2016 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <ldb.h>
+#include <talloc.h>
+#include <tevent.h>
+
+#include "util/util.h"
+#include "responder/common/cache_req/cache_req_private.h"
+#include "responder/common/cache_req/cache_req_plugin.h"
+
+static struct cache_req_plugin *
+cache_req_get_plugin(enum cache_req_type type)
+{
+    static struct cache_req_plugin *plugins[CACHE_REQ_SENTINEL] = {
+
+    };
+
+    if (type >= CACHE_REQ_SENTINEL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: invalid plugin type!");
+        return NULL;
+    }
+
+    return plugins[type];
+}
+
+static void cache_req_set_plugin(struct cache_req *cr,
+                                 enum cache_req_type type)
+{
+    struct cache_req_plugin *plugin;
+
+    plugin = cache_req_get_plugin(type);
+    if (plugin == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Bug: unset plugin!");
```

Comment:
Not really. If the plugin is missing then there is serious error in the code. But I added a check so we don't crash if such mistake ever got in production.